include/ginkgo/core/preconditioner/ilu.hpp

```diff
@@ -0,0 +1,407 @@
+     * `with_criteria`.
+     *
+     * Contains a constexpr boolean `value`, which is true if the Factory class
+     * of SolverType has a `with_criteria`, and false otherwise.
+     *
+     * @tparam SolverType   Solver to test if its factory has a with_criteria
+     *                      function.
+     *
+     */
+    template <typename SolverType, typename = void>
+    struct has_with_criteria : std::false_type {};
+
+    /**
+     * @copydoc has_with_criteria
+     *
+     * @internal  The second template parameter (which uses SFINAE) must match
+     *            the default value of the general case in order to be accepted
+     *            as a specialization.
+     */
+    template <typename SolverType>
+    struct has_with_criteria<
+        SolverType,
+        xstd::void_t<decltype(std::declval<factory_type_t<SolverType>>()
+                                  .with_criteria(with_criteria_param_type()))>>
+        : std::true_type {};
+
+
+    /**
+     * Generates a default solver of type SolverType.
+     *
+     * Also checks whether SolverType can be assigned a criteria, and if it
+     * can, it is assigned default values which should be well suited for a
+     * preconditioner.
+     */
+    template <typename SolverType>
+    static xstd::enable_if_t<has_with_criteria<SolverType>::value,
+                             std::unique_ptr<SolverType>>
+    generate_default_solver(const std::shared_ptr<const Executor> &exec,
+                            const std::shared_ptr<const LinOp> &mtx)
+    {
+        constexpr value_type default_reduce_residual{1e-4};
+        const unsigned int default_max_iters{
+            static_cast<unsigned int>(mtx->get_size()[0])};
+
+        return SolverType::build()
+            .with_criteria(gko::stop::Iteration::build()
+                               .with_max_iters(default_max_iters)
+                               .on(exec),
+                           gko::stop::ResidualNormReduction<>::build()
+                               .with_reduction_factor(default_reduce_residual)
+                               .on(exec))
+            .on(exec)
+            ->generate(mtx);
+    }
+
+    /**
+     * @copydoc generate_default_solver
+     */
+    template <typename SolverType>
+    static xstd::enable_if_t<!has_with_criteria<SolverType>::value,
+                             std::unique_ptr<SolverType>>
+    generate_default_solver(const std::shared_ptr<const Executor> &exec,
+                            const std::shared_ptr<const LinOp> &mtx)
+    {
+        return SolverType::build().on(exec)->generate(mtx);
+    }
+
+private:
+    std::shared_ptr<const l_solver_type> l_solver_{};
+    std::shared_ptr<const u_solver_type> u_solver_{};
+    /**
+     * Manages a vector as a cache, so there is no need to allocate one every
+     * time an intermediate vector is required.
+     * Copying an instance will only yield an empty object since copying the
```

Comment:
if you dont want it to be copied, you could delete the copy and move constructors.

Reply:
No, I can't, every PolymorphicObject needs to be copyable (see definition for [`copy_from`](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/polymorphic_object.hpp#L157)). In the end, this is called in [EnablePolymorphicAssignment](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/polymorphic_object.hpp#L616).
If it is not, it simply does not compile.

Reply:
Hm, you are right.